### PR TITLE
DO NOT SQUASH: Test infinite model with ABC solvers (5)

### PR DESCRIPTION
**Pull request by ruansava**
This PR tests the implementation of a reference model for ABC schemes based on layers:
1. Hybrid
2. PML

**Absorbing Boundary Condition (ABC) Handling:**
* Refined ABC logic in the main solver workflow to ensure correct initialization based on the selected ABC type (`AbsorbingBCsType.NOABCS` and `AbsorbingBCsType.NRBC` now trigger model parameter initialization).
* Improved PML layer setup to check and assign the velocity model if missing, and cleaned up the computation of maximum damping coefficients and mesh coordinates.

**Error Checking and Validation Enhancements:**
* Extended `type_data_structure_error` to support checks for 2D and 3D arrays, including optional shape validation, and improved type element checking for nested structures.
* Added support for checking `Form` objects in `type_firedrake_error` and made type validation more robust.

**Documentation and Comments:**
* Improved and clarified docstrings and comments in several files, including fixing typos and making parameter descriptions more precise.
* Updated comments and variable names for better clarity, e.g., changing "damping type" to "ABC type".

### CODE REVIEW

**Review by Olender — CHANGES_REQUESTED** [review]
This 'request changes' is just to stop any merge to main (github is partially down, they even posted in https://www.githubstatus.com/) and its doing weird things to our runners and backup tag generation

File: spyro/abc/abc_layer.py [hunk 1]
```diff
@@ -848,210 +841,236 @@ def nrbc_on_boundary_layer(self, wave, non_reflect_bc, save_file=True):
...
+        # Maximum timestep size
+        dt_sol = Modal_Solver(self.dimension, method=method, calc_max_dt=True)
+        max_dt = dt_sol.estimate_timestep(wave.c, wave.function_space, wave.final_time,
+                                          shift=1e-8, quad_rule=wave.quadrature_rule,
+                                          fraction=1.)
+
+        # Rounding power
+        pot = int(abs(ceil(log10(max_dt))) + mag_add)
+
+        # Maximum timestep size according to divisors of the final time
```
Comment (Olender) [inline spyro/abc/abc_layer.py:894]: maybe this could be a helper function in the future

Comment (ruansava) [reply spyro/abc/abc_layer.py:894]: Ok! If you provide me an example of generalization, I can transform it into a helper

File: spyro/abc/abc_layer.py [hunk 2]
```diff
@@ -848,210 +841,236 @@ def nrbc_on_boundary_layer(self, wave, non_reflect_bc, save_file=True):
...
+
+        str_pad = "Infinite Domain Extension Based on "
+
+        # Distance already travelled by the wave
+        if hasattr(self, 'eik_bnd'):
+
+            str_pad += "Minimun Eikonal at Critical Boundary Points"
+
+            # If Eikonal analysis was performed
+            eikmin = self.eik_bnd[0][2]
```
Comment (Olender) [inline spyro/abc/abc_layer.py:960]: The 2 index is the same in 2D and 3D cases?

File: spyro/abc/abc_layer.py [hunk 3]
```diff
@@ -848,210 +841,236 @@ def nrbc_on_boundary_layer(self, wave, non_reflect_bc, save_file=True):
...
+
+            # Minimum distance to the nearest boundary
+            dist_to_bnd = c_bnd_max * eikmin / 2.
+        else:
+
+            str_pad += "Minimum Distance Source-Boundary"
+
+            # If Eikonal analysis was not performed
+            sou_loc = array(source_locations)
```
Comment (Olender) [inline spyro/abc/abc_layer.py]: what about source_locations_array or source_loc or anything more descriptive?

File: spyro/pml/pml_nsnc.py [hunk 4]
```diff
@@ -266,6 +266,11 @@ def pml_layer(self, wave, save_file=True):
         None
         """
 
+        # Check if the velocity model exists
+        # TODO: Delete support for initial_velocity_model
+        if wave.c is None:
```
Comment (Olender) [inline spyro/pml/pml_nsnc.py:271]: Maybe this bug goes away with PR#356 ?

File: spyro/tools/error_measure.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+    Methods
+    -------
+    error_measures_habc()
+        Compute the error measures at the receivers for the HABC scheme
+    get_reference_signal()
+        Acquire the reference signal to compare with the HABC scheme
+    save_reference_signal()
+        Save the reference signal for the HABC scheme
+
+    comparison_plots()
```
Comment (Olender) [inline spyro/tools/error_measure.py:41]:
```suggestion
        Save the reference signal for the HABC scheme
    comparison_plots()
```

File: spyro/abc/abc_layer.py (hunk 3 above)
Comment (ruansava) [reply spyro/abc/abc_layer.py]: Since it comes to source_locations and the method is short I did not see the needed to assign a large name, but I can add more letters 🙃

File: spyro/abc/abc_layer.py (hunk 2 above)
Comment (ruansava) [reply spyro/abc/abc_layer.py:960]: Yes, this is a list with specific data provided in the documentation. Also, I will add the documentation here.

File: spyro/utils/error_management.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 from numpy import float32, float64, inf, int32, int64, isinf, isnan, ndarray
 from firedrake import Function, FunctionSpace, Mesh
 from firedrake.functionspaceimpl import WithGeometry
+from ufl.form import Form
 from ufl.geometry import SpatialCoordinate
```
Comment (Olender) [inline spyro/utils/error_management.py:11]:
```suggestion
from ufl.geometry import SpatialCoordinate
from os import path


def value_file_error(
        parameter_name,
        parameter_value,
        valid_extensions,
        accept_parameter_as_none=False,
        check_file_existance=False,
    ):
    """Validate a file parameter and raise a ValueError if invalid.

    Parameters
    ----------
    parameter_name : `str`
        Name of the parameter to be validated (used in error messages).
    parameter_value : `str`
        Value of the parameter to be validated.
    valid_extensions : `list`
        List of valid file extensions for the parameter.
    accept_parameter_as_none : `bool`, optional
        If `True`, the parameter value is allowed to be validated as `None`.
        Default is `False`, in which case `None` is not allowed.
    check_file_existance : `bool`, optional
        If `True`, we will check if the file existis.
        Default is `False`

    Returns
    -------
    par_value : `str`
        The validated file path or file name.

    Raises
    ------
    TypeError
        If the parameter value is not a `str`.
    ValueError
        If the file extension is not in the list of valid extensions.
    FileNotFoundError
        If the file does not exist and check_exists is True
    """

    parameter_value = value_string_error(parameter_name, parameter_value, none_default=accept_parameter_as_none)

    if not path.exists(parameter_value) and check_file_existance:
        raise FileNotFoundError(f"'{parameter_name}' '{parameter_value}' does not exist")

    file_extension = path.splitext(parameter_value)[1].lower()
    value_parameter_error("extension_type", file_extension, valid_extensions)

    return parameter_value
```

File: spyro/tools/error_measure.py [hunk 7]
```diff
@@ -0,0 +1,548 @@
+from os import getcwd
+from numpy import concatenate, inf, load, save, savetxt, trapezoid, zeros
+from scipy.signal import find_peaks
+from ..io.basicio import parallel_print as pprint
+from ..utils.error_management import (type_data_structure_error, value_numerical_error,
+                                      value_string_error)
```
Comment (Olender) [inline spyro/tools/error_measure.py]:
```suggestion
from ..utils.error_management import (
    type_data_structure_error,
    value_numerical_error,
    value_string_error,
    value_file_error,
)
```

File: spyro/tools/error_measure.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+        pprint("\nLoading Reference Signal from Reference Model", comm=self.comm)
+
+        # Path to the reference data folder with reference signals
+        pth_str = self.path_reference + self.output_file + "_"
+
+        # Time domain signal
+        receivers_reference = load(pth_str + "time.npy")
+
+        # Frequency domain signal
+        receivers_ref_fft = load(pth_str + "fft.npy").T
```
Comment (Olender) [inline spyro/tools/error_measure.py]:
```suggestion
        # Path to the reference data folder with reference signals
        pth_str = self.path_reference + self.output_file + "_"

        # Time domain signal
        receivers_reference_file = value_file_error(pth_str + "time.npy", check_file_existance=True)
        receivers_reference = load(receivers_reference_file)

        # Frequency domain signal
        receivers_reference_fft_file = value_file_error(pth_str + "fft.npy", check_file_existance=True)
        receivers_ref_fft = load(receivers_reference_fft_file).T
```

File: spyro/pml/pml_nsnc.py (hunk 4 above)
Comment (ruansava) [reply spyro/pml/pml_nsnc.py:271]: In fact, by applying the solution in issue #348; and forcing to add a pad > 0 if ABCs based on layers are required. Thus, the default solver is used.

File: spyro/utils/error_management.py (hunk 6 above)
Comment (ruansava) [reply spyro/utils/error_management.py:11]: Do you want I include tests for this?
I am going to adapt according to the current structure. For example, it `None` is default then `none_default`=False.

File: spyro/abc/abc_layer.py (hunk 3 above)
Comment (Olender) [reply spyro/abc/abc_layer.py]: Perfect! More letters please! sou_loc is always autocorrected in my brain to 'sou louco', which, to be fair, is the natural progression of working with spiro for too long. I speak from experience

File: spyro/tools/error_measure.py (hunk 5 above)
Comment (ruansava) [reply spyro/tools/error_measure.py:41]: Ok

File: spyro/tools/error_measure.py (hunk 8 above)
Comment (ruansava) [reply spyro/tools/error_measure.py]: I liked this

**Review by ruansava — COMMENTED** [review]
I will add tests for the new `value_file_error` method

File: spyro/utils/error_management.py (hunk 6 above)
Comment (Olender) [reply spyro/utils/error_management.py:11]: But the method doesn't need to know if None is or not a default outside of it, just if we want it to accept None

Comment (Olender) [reply spyro/utils/error_management.py:11]: > Do you want I include tests for this? I am going to adapt according to the current structure. For example, it `None` is default then `none_default`=False.

Just don't delete my characters! I really like parameter_value instead of par_value, since I made this method specifically

Comment (ruansava) [reply spyro/utils/error_management.py:11]: Too late! :v
But feel free to change all the names in the next PR dedicated to validation if they make more sense.

**Comment by ruansava** [comment]
Before merging, I'm going to try to further reduce the test time for the infinite 3D model. So far, the reduction has been from 28 min to 13.5 min and then to 8 min per test. The goal is 5 min as the maximum expected time.

File: tests/on_one_core/test_analytical_solution.py [hunk 9]
```diff
@@ -53,6 +53,13 @@ def test_analytical_solution(use_vertex_only_mesh):
 
     nt = len(time_vector)
     error = error_calc(numerical_p, analytical_p, nt)
+
+    # import ipdb
+    # ipdb.set_trace()
+
+    # errPk = MeasureError().peak_error(numerical_p, analytical_p)[0]
+    # errIt = MeasureError().integral_error(numerical_p, analytical_p, wave.dt)
```
Comment (Olender) [inline tests/on_one_core/test_analytical_solution.py]: Why the added commented lines? We could just open an issue for this in another PR

Comment (ruansava) [reply tests/on_one_core/test_analytical_solution.py]: I forgot to delete these lines. It is something that I am going to do in another PR

**Comment by ruansava** [comment]
> Before merging, I'm going to try to further reduce the test time for the infinite 3D model. So far, the reduction has been from 28 min to 13.5 min and then to 8 min per test. The goal is 5 min as the maximum expected time.

I reached the goal of 5min. I’ll mark as slow tests those that take more than 200s.

**Review by Olender — APPROVED** [review]
